compiler/rustc_resolve/src/imports.rs

```diff
@@ -338,7 +338,15 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
                         } else {
                             resolution.binding = Some(nonglob_binding);
                         }
-                        resolution.shadowed_glob = Some(glob_binding);
+
+                        if key.ns == TypeNS
+                            && resolution
+                                .shadowed_glob
+                                .map_or(false, |old| old.res() != glob_binding.res())
+                        {
+                        } else {
+                            resolution.shadowed_glob = Some(glob_binding);
+                        }
```

Comment:
I think it would be sufficient to not set shadowed glob if it was already set previously?
```rust
if resolution.shadowed_glob.is_none() { resolution.shadowed_glob = Some(glob_binding); }
```

Reply:
What would be even more correct is to treat to competing shadowed globs as two competing regular globs, and do this thing
```rust
                match (old_binding.is_glob_import(), binding.is_glob_import()) {
                    (true, true) => {
                        if res != old_binding.res() {
                            resolution.binding = Some(this.ambiguity(
                                AmbiguityKind::GlobVsGlob,
                                old_binding,
                                binding,
                            ));
                        } else if !old_binding.vis.is_at_least(binding.vis, this.tcx) {
                            // We are glob-importing the same item but with greater visibility.
                            resolution.binding = Some(binding);
                        }
                    }
```
but with `resolution.shadowed_glob` instead of `resolution.binding`.

Reply:
It would be compatible with the proper solution

https://github.com/rust-lang/rust/blob/master/compiler/rustc_resolve/src/ident.rs#L874-L876

which requires a larger refactoring.